<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Large-Scale Allele-Specific CNV Detection Using Coverage and Allelic Imbalance in WES data • XploR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Large-Scale Allele-Specific CNV Detection Using Coverage and Allelic Imbalance in WES data">
<meta name="description" content="Detects large-scale, allele-specific copy number variations (CNVs) from tumor-only sequencing data. Leverages both coverage depth and allelic imbalance information to infer genome-wide copy number states and allelic configurations. Designed for whole-exome sequencing. Suitable for detecting aneuploidy, large deletions or gains, and loss of heterozygosity (LOH) in cancer genomics applications.">
<meta property="og:description" content="Detects large-scale, allele-specific copy number variations (CNVs) from tumor-only sequencing data. Leverages both coverage depth and allelic imbalance information to infer genome-wide copy number states and allelic configurations. Designed for whole-exome sequencing. Suitable for detecting aneuploidy, large deletions or gains, and loss of heterozygosity (LOH) in cancer genomics applications.">
<meta property="og:image" content="https://sj-cmpb-se.github.io/XploR/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">XploR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sj-cmpb-se/XploR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p align="center">
<img src="reference/figures/logo.png" height="150"></p>
<p><strong>XploR</strong> is an R package for robust, large-scale copy number and allelic imbalance analysis from whole exome sequencing (WES) data.<br>
It provides tools for segmentation, purity/ploidy estimation, allelic imbalance analysis, ISCN annotation, interactive plotting, and more.</p>
<hr>
<div class="section level2">
<h2 id="features">Features<a class="anchor" aria-label="anchor" href="#features"></a>
</h2>
<ul>
<li>Exome-wide copy number segmentation and allelic imbalance detection</li>
<li>Purity and ploidy estimation with model selection</li>
<li>BAF and coverage smoothing, binning, and quality control</li>
<li>ISCN and gene annotation of CNV segments</li>
<li>Interactive and publication-ready visualization</li>
<li>Batch and command-line workflows for high-throughput analysis</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the latest version from GitHub using <a href="https://github.com/r-lib/devtools" class="external-link">devtools</a>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"sj-cmpb-se/XploR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>This is a basic example to run the program:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sj-cmpb-se.github.io/XploR/">XploR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Example: Run main CNV workflow (see vignette for details)</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/runAIsegmentation.html">runAIsegmentation</a></span><span class="op">(</span></span>
<span> seg <span class="op">=</span> <span class="st">"sample.seg"</span>, </span>
<span>  cov <span class="op">=</span> <span class="st">"sample.counts"</span>,</span>
<span>  ai <span class="op">=</span> <span class="st">"sample.tumor.baf.bedgraph.gz"</span>,</span>
<span>  gender <span class="op">=</span> <span class="st">"female"</span>,</span>
<span>  out_dir <span class="op">=</span> <span class="st">"results/"</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">"Sample1"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Annotate segments with ISCN and gene information</span></span>
<span><span class="fu"><a href="reference/AnnotateSegments.html">AnnotateSegments</a></span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="st">"results/Sample1_GATK_DRAGEN_merge.tsv"</span>,</span>
<span>  out_dir <span class="op">=</span> <span class="st">"results/"</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">"Sample1"</span>,</span>
<span>  cytoband <span class="op">=</span> <span class="st">"data/cytoBand.txt"</span>,</span>
<span>  whitelist_edge <span class="op">=</span> <span class="st">"data/whitelist.txt"</span>,</span>
<span>  gene <span class="op">=</span> <span class="st">"data/gene_anno.txt"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot results</span></span>
<span><span class="fu">plot_cnv_profile</span><span class="op">(</span></span>
<span>  seg <span class="op">=</span> <span class="st">"results/Sample1_CNV_annotation.tsv"</span>,</span>
<span>  coverage <span class="op">=</span> <span class="st">"sample.denoisedCR.tsv"</span>,</span>
<span>  ballele <span class="op">=</span> <span class="st">"sample.tumor.ballele.counts.gz"</span>,</span>
<span>  whitelist <span class="op">=</span> <span class="st">"data/whitelist.txt"</span>,</span>
<span>  gender <span class="op">=</span> <span class="st">"female"</span>,</span>
<span>  out_dir <span class="op">=</span> <span class="st">"results/"</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">"Sample1"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/sj-cmpb-se/XploR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/sj-cmpb-se/XploR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing XploR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Meiling Jin <br><small class="roles"> Maintainer </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Meiling Jin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
