<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • XploR</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://sj-cmpb-se.github.io/XploR/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">XploR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sj-cmpb-se/XploR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="AddQualTag.html">AddQualTag()</a></code>

        </dt>
        <dd>Assign Quality Tag to a Segment Based on BAF GMM Metrics</dd>
      </dl><dl><dt>

          <code><a href="AnnotateSegments.html">AnnotateSegments()</a></code>

        </dt>
        <dd>Annotate CNV Segments with ISCN and Gene Information</dd>
      </dl><dl><dt>

          <code><a href="AssignPriors.html">AssignPriors()</a></code>

        </dt>
        <dd>Assign Priors to Major/Minor Allele Combinations Based on Biological Difficulty</dd>
      </dl><dl><dt>

          <code><a href="CalMedian.html">CalMedian()</a></code>

        </dt>
        <dd>Calculate Median Coverage for a Genomic Region</dd>
      </dl><dl><dt>

          <code><a href="CalSM.html">CalSM()</a></code>

        </dt>
        <dd>Calculate Segment Mean Adjusted by Scale Factor</dd>
      </dl><dl><dt>

          <code><a href="CalSegmentLikelihood.html">CalSegmentLikelihood()</a></code>

        </dt>
        <dd>Calculate Marginal Likelihood for a Single Segment</dd>
      </dl><dl><dt>

          <code><a href="CalbaselineCov.html">CalbaselineCov()</a></code>

        </dt>
        <dd>Calculate Baseline Coverage</dd>
      </dl><dl><dt>

          <code><a href="Calccf.html">Calccf()</a></code>

        </dt>
        <dd>Calculate Cancer Cell Fraction (CCF)</dd>
      </dl><dl><dt>

          <code><a href="CallMerge.html">CallMerge()</a></code>

        </dt>
        <dd>Merge Segments or AI Rows by Chromosome</dd>
      </dl><dl><dt>

          <code><a href="CallWTModel.html">CallWTModel()</a></code>

        </dt>
        <dd>Assign Copy Number Call With Model (Major/Minor)</dd>
      </dl><dl><dt>

          <code><a href="Callikelihood.html">Callikelihood()</a></code>

        </dt>
        <dd>Calculate Likelihood for Each \(\mu\) and \(\rho\) Combination</dd>
      </dl><dl><dt>

          <code><a href="CallwoModel.html">CallwoModel()</a></code>

        </dt>
        <dd>Assign Copy Number Call Without Model</dd>
      </dl><dl><dt>

          <code><a href="CcfLOH.html">CcfLOH()</a></code>

        </dt>
        <dd>Calculate Cancer Cell Fraction (CCF) from Allelic Imbalance (LOH)</dd>
      </dl><dl><dt>

          <code><a href="CheckGender.html">CheckGender()</a></code>

        </dt>
        <dd>Check and Adjust Gender in GATK Segmentation Data</dd>
      </dl><dl><dt>

          <code><a href="Checkmode.html">Checkmode()</a></code>

        </dt>
        <dd>Check and Set Mode-Specific Options</dd>
      </dl><dl><dt>

          <code><a href="ClusterAdjacent.html">ClusterAdjacent()</a></code>

        </dt>
        <dd>Cluster Adjacent Values Based on a Threshold</dd>
      </dl><dl><dt>

          <code><a href="ClusterModels.html">ClusterModels()</a></code>

        </dt>
        <dd>Cluster and Select the Best Models Based on Likelihood and Copy Number Distance</dd>
      </dl><dl><dt>

          <code><a href="CorrectBAF.html">CorrectBAF()</a></code>

        </dt>
        <dd>Correct BAF Values Using a Scaling Factor</dd>
      </dl><dl><dt>

          <code><a href="CorrectGender.html">CorrectGender()</a></code>

        </dt>
        <dd>Correct Copy Ratio File if GATK and Pipeline Gender Disagree</dd>
      </dl><dl><dt>

          <code><a href="CorrectPurity.html">CorrectPurity()</a></code>

        </dt>
        <dd>Correct CNF and MAF According to Purity</dd>
      </dl><dl><dt>

          <code><a href="CoverageQC.html">CoverageQC()</a></code>

        </dt>
        <dd>Summarise BAF QC from CNV Segment Annotation</dd>
      </dl><dl><dt>

          <code><a href="EstimateBAFbyGMM.html">EstimateBAFbyGMM()</a></code>

        </dt>
        <dd>Estimate BAF Cluster Mean and Weight Using Gaussian Mixture Modeling</dd>
      </dl><dl><dt>

          <code><a href="EstimateBAFshift.html">EstimateBAFshift()</a></code>

        </dt>
        <dd>Estimate BAF Shift Scaling Factor</dd>
      </dl><dl><dt>

          <code><a href="EstimateCovSF.html">EstimateCovSF()</a></code>

        </dt>
        <dd>Estimate Coverage Scale Factor for a Chromosome or Region</dd>
      </dl><dl><dt>

          <code><a href="EstimateMinPurity.html">EstimateMinPurity()</a></code>

        </dt>
        <dd>Estimate Minimum Tumor Purity from BAF Data</dd>
      </dl><dl><dt>

          <code><a href="EstimatePurityCov.html">EstimatePurityCov()</a></code>

        </dt>
        <dd>Estimate Tumor Purity from Coverage and Segmentation Data</dd>
      </dl><dl><dt>

          <code><a href="EstimateSFVariance.html">EstimateSFVariance()</a></code>

        </dt>
        <dd>Estimate Standard Deviation of a Numeric Vector</dd>
      </dl><dl><dt>

          <code><a href="ExtractCall.html">ExtractCall()</a></code>

        </dt>
        <dd>Extract Final Calls for a Given (mu, rho) Model</dd>
      </dl><dl><dt>

          <code><a href="FindTier1Models.html">FindTier1Models()</a></code>

        </dt>
        <dd>Find the "Elbow" Point in a Ranked Vector (Tier 1 Model Selection)</dd>
      </dl><dl><dt>

          <code><a href="FixsegmentMean.html">FixsegmentMean()</a></code>

        </dt>
        <dd>Fix Segment Mean Based on Gender Discrepancy</dd>
      </dl><dl><dt>

          <code><a href="GenerateCombinations.html">GenerateCombinations()</a></code>

        </dt>
        <dd>Generate All Major/Minor Allele Combinations for a Given Total Copy Number</dd>
      </dl><dl><dt>

          <code><a href="Groupvalues.html">Groupvalues()</a></code>

        </dt>
        <dd>Group Values by Change Points in the Mean</dd>
      </dl><dl><dt>

          <code><a href="MergeAICheck.html">MergeAICheck()</a></code>

        </dt>
        <dd>Check Whether Two AI Segments Should Be Merged</dd>
      </dl><dl><dt>

          <code><a href="MergeAIRow.html">MergeAIRow()</a></code>

        </dt>
        <dd>Merge Adjacent AI Segments Based on Similarity Criteria</dd>
      </dl><dl><dt>

          <code><a href="MergeSegCheck.html">MergeSegCheck()</a></code>

        </dt>
        <dd>Check Whether Two Segments Should Be Merged (Copy Number)</dd>
      </dl><dl><dt>

          <code><a href="MergeSegRow.html">MergeSegRow()</a></code>

        </dt>
        <dd>Merge Adjacent Copy Number Segments Based on Similarity Criteria</dd>
      </dl><dl><dt>

          <code><a href="ModelSource.html">ModelSource()</a></code>

        </dt>
        <dd>Determine Model Source Based on BAF and Coverage Information</dd>
      </dl><dl><dt>

          <code><a href="ParseParm.html">ParseParm()</a></code>

        </dt>
        <dd>Parse Purity and Scale Factor Parameters for Model Selection</dd>
      </dl><dl><dt>

          <code><a href="PlotCovDisCN.html">PlotCovDisCN()</a></code>

        </dt>
        <dd>Plot Coverage Distance to Integer Copy Number vs. Purity</dd>
      </dl><dl><dt>

          <code><a href="PlotLikelihoodAndDistance.html">PlotLikelihoodAndDistance()</a></code>

        </dt>
        <dd>Plot Model Likelihood and Distance Clusters as an Interactive Heatmap</dd>
      </dl><dl><dt>

          <code><a href="PlotModel.html">PlotModel()</a></code>

        </dt>
        <dd>Plot Model Total Log-Likelihoods as a Heatmap</dd>
      </dl><dl><dt>

          <code><a href="PlotModelDot.html">PlotModelDot()</a></code>

        </dt>
        <dd>Plot Model Total Log-Likelihoods as a Dot Plot</dd>
      </dl><dl><dt>

          <code><a href="RefineCalls.html">RefineCalls()</a></code>

        </dt>
        <dd>Refine Segment Calls According to the Final Model</dd>
      </dl><dl><dt>

          <code><a href="RefineCallsSecond.html">RefineCallsSecond()</a></code>

        </dt>
        <dd>Second Refinement of Segment Calls for Cov/AI Profile Mismatches</dd>
      </dl><dl><dt>

          <code><a href="RefineTier1Models.html">RefineTier1Models()</a></code>

        </dt>
        <dd>Refine Tier 1 Models and Their Segment Calls</dd>
      </dl><dl><dt>

          <code><a href="RunCallikelihood.html">RunCallikelihood()</a></code>

        </dt>
        <dd>Run Likelihood Calculation Across All Purity/Mu Combinations</dd>
      </dl><dl><dt>

          <code><a href="SearchBreakpoint.html">SearchBreakpoint()</a></code>

        </dt>
        <dd>Search for Breakpoints within a Segment Using BAF Data</dd>
      </dl><dl><dt>

          <code><a href="SegmentMeanToOriCov.html">SegmentMeanToOriCov()</a></code>

        </dt>
        <dd>Convert Segment Mean to Original Coverage</dd>
      </dl><dl><dt>

          <code><a href="SelectCallpersegment.html">SelectCallpersegment()</a></code>

        </dt>
        <dd>Select the Most Likely Call per Segment</dd>
      </dl><dl><dt>

          <code><a href="SelectFinalModel.html">SelectFinalModel()</a></code>

        </dt>
        <dd>Select the Final Copy Number Model and Write Results</dd>
      </dl><dl><dt>

          <code><a href="SelectModelByDis.html">SelectModelByDis()</a></code>

        </dt>
        <dd>Select Model by Distance to Integer Copy Number</dd>
      </dl><dl><dt>

          <code><a href="SummarizeMapQC.html">SummarizeMapQC()</a></code>

        </dt>
        <dd>Summarise QC Tables for a Sample</dd>
      </dl><dl><dt>

          <code><a href="ai_bin_smooth.html">ai_bin_smooth()</a></code>

        </dt>
        <dd>Smooth AI Bin Values by Downsampling to Mode Proportions</dd>
      </dl><dl><dt>

          <code><a href="band_plottable_female.html">band_plottable_female</a></code>

        </dt>
        <dd>Non blacklist region for each cytobands binned by 300nt for female.</dd>
      </dl><dl><dt>

          <code><a href="band_plottable_male.html">band_plottable_male</a></code>

        </dt>
        <dd>Non blacklist region for each cytobands binned by 300nt for male.</dd>
      </dl><dl><dt>

          <code><a href="binbaf.html">binbaf()</a></code>

        </dt>
        <dd>Bin BAF Data and Estimate BAF Metrics per Bin</dd>
      </dl><dl><dt>

          <code><a href="blacklist_table.html">blacklist_table</a></code>

        </dt>
        <dd>Non mapable regions of the genome</dd>
      </dl><dl><dt>

          <code><a href="categorize_K.html">categorize_K()</a></code>

        </dt>
        <dd>Categorize Beta Distribution K Values</dd>
      </dl><dl><dt>

          <code><a href="clean_homalt.html">clean_homalt()</a></code>

        </dt>
        <dd>Clean Homozygous AI Bins Based on LOH Segments</dd>
      </dl><dl><dt>

          <code><a href="correct_purity.html">correct_purity()</a></code>

        </dt>
        <dd>Correct Copy Ratio and Calculate CNF for Purity and Scale Factor</dd>
      </dl><dl><dt>

          <code><a href="cov_bin.html">cov_bin()</a></code>

        </dt>
        <dd>Bin and Smooth Copy Number Data from Coverage File</dd>
      </dl><dl><dt>

          <code><a href="cytoband_table.html">cytoband_table</a></code>

        </dt>
        <dd>Cytoband reference table</dd>
      </dl><dl><dt>

          <code><a href="edge_table.html">edge_table</a></code>

        </dt>
        <dd>Detectable Edge table for cytoband annotation follows the 5MB rule.</dd>
      </dl><dl><dt>

          <code><a href="gene_annotation.html">gene_annotation</a></code>

        </dt>
        <dd>RefSeq gene annotations</dd>
      </dl><dl><dt>

          <code><a href="keep_whitelist_ai.html">keep_whitelist_ai()</a></code>

        </dt>
        <dd>Keep Only AI Bins in Whitelisted Regions</dd>
      </dl><dl><dt>

          <code><a href="keep_whitelist_cov.html">keep_whitelist_cov()</a></code>

        </dt>
        <dd>Keep Only Coverage Bins in Whitelisted Regions</dd>
      </dl><dl><dt>

          <code><a href="plot_cov.html">plot_cov()</a></code>

        </dt>
        <dd>Multi-panel Plot of Copy Number, BAF, Clonality, and Quality Across Chromosomes</dd>
      </dl><dl><dt>

          <code><a href="read_whitelist.html">read_whitelist()</a></code>

        </dt>
        <dd>Read and Process DRAGEN Whitelist Regions</dd>
      </dl><dl><dt>

          <code><a href="rerunCalling.html">rerunCalling()</a></code>

        </dt>
        <dd>Rerun CNV Calling with User-Defined Purity and Size Factor</dd>
      </dl><dl><dt>

          <code><a href="round_ai.html">round_ai()</a></code>

        </dt>
        <dd>Round BAF Values into 20 Bins</dd>
      </dl><dl><dt>

          <code><a href="roundcn.html">roundcn()</a></code>

        </dt>
        <dd>Round Copy Number to Integer Based on Call and Gender</dd>
      </dl><dl><dt>

          <code><a href="runAIsegmentation.html">runAIsegmentation()</a></code>

        </dt>
        <dd>Run AI Segmentation Workflow</dd>
      </dl><dl><dt>

          <code><a href="runModelLikelihood.html">runModelLikelihood()</a></code>

        </dt>
        <dd>Run Model Likelihood and Copy Number Calling Workflow</dd>
      </dl><dl><dt>

          <code><a href="runPlotCNV.html">runPlotCNV()</a></code>

        </dt>
        <dd>Plot CNV Profile (Copy Number, BAF, Clonality, Quality)</dd>
      </dl><dl><dt>

          <code><a href="smooth_ai.html">smooth_ai()</a></code>

        </dt>
        <dd>Smooth AI Bins by Window Size and Downsampling</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Meiling Jin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

